### ISSUE TITLE
Wallclock time and capping

### ISSUE THREAD

**Issue by leobezerra** [automl/auto-sklearn#321]
Hi guys.

I was running some trial experiments yesterday after installing Auto-Sklearn and I noticed a couple things that are probably not working as they should:

1. I started a run around 16h15 using default parameters (1h run), and had to shut my laptop around 16h25. By the time I woke my laptop up around 17h15, it resumed running but concluded shortly after. Are you timing the wallclock time or the CPU time? It would be nice to have a flag control this.

2. I decided to run again Auto-Sklearn to make sure I used 1h. The issue was that the accuracy I got from the 1h run was the same from the 10-15 min run. Isn't there some sort of capping rule stating that if, after a given number of sampled candidates or elapsed runtime, the incumbent is not improving, the run should be capped?

To provide a few more details, I used the default parameter settings for both runs, the dataset was MNIST, and the accuracy score was 0.9593.

Thanks in advance.

**Comment by mfeurer** [comment]
Hi @leobezerra. Thank you very much for the feedback. First of all, in order to get to a result of ~98.5 accuracy on MNIST, you need to run auto-sklearn for one or two days. We're working on improvements on this, but that's the current status. Regarding 

1. Our current assumption is that the user has his/her machine constantly running. In order to measure CPU time, we would have to measure CPU time of the auto-sklearn parent process and the child processes which execute the machine learning algorithm. We start the child processes with a tool called pynisher. As far as I know it's not possible to measure the CPU time with this (@sfalkner could you please let me know if my answer is correct). Therefore, I'm afraid you need to leave your machine on or use auto-sklearn on a machine which is constantly on.
2. The optimization process can appear to stall for quite a while. However, keep in mind that we cannot asses whether we have found the optimal combination of hyperparameters and therefore this is a non-trivial problem. I am not familiar with any capping rule that is scientifically founded on when to stop Bayesian optimization if no improvement is made. Therefore, the time to run auto-sklearn is unfortunately a hyperparameter of auto-sklearn.

I hope this helps.

I hope this helps.

**Comment by sfalkner** [comment]
Hey,
the pynisher can limit the CPU time of the subprocess, but it doesn't measure it and things can get hairy once the ML algorithm spawns even more subprocesses and if multiple cores are used. AFAIK, all childer count towards their parent process' CPU time. So limiting that in auto sklearn to handle suspending the computation is not trivial. Usually, you want to run it for an extended period anyway, so maybe you can run it over night. Sorry that I don't have better news.
Best,
Stefan

**Comment by leobezerra** [comment]
Hi guys.

About the wallclock time, if that's it, that's it 👍 
The issue is that it becomes very difficult to compare things fairly, since 1d in your setup is probably very different from 1d in one of my setups. Would it be possible to at least have some log info on how many configurations were tested? Another option is precisely that: to limit the number of configurations tested. (mimics the maximum FE stopping criterion from search optimization)

About the search getting stalled, I know it's not common practice capping runs in machine learning, but in optimization we do that almost as default.

To me, it's quite impressive that you say that, increasing from 10min to 1h doesn't make any difference in accuracy, but from 1h to 1d it does.

I'll try to launch an auto-sklearn job on my university computer for a day and see what I get.

**Comment by mfeurer** [comment]
You can check the attribute `cv_results_` of the [AutoSklearnClassifier](http://automl.github.io/auto-sklearn/stable/api.html) (and the regressor) for the number of function evaluations. 

I think the issue why there is no improvement from 10 minutes to 1h comes from the fact that a lot of algorithms take a few minutes on MNIST. The 95% accuracy which you observe is the accuracy of the default configuration. After that, auto-sklearn will start with a set of configurations found by meta-learning and then start exploring the space. I guess that a lot of configurations from meta-learning hit the default time limit, which might be another reason why you don't observe any improvements. You could try to set the time limit of each individual ML algorithm to ~30 minutes.

Capping runs early is on our TODO list, but we don't expect to have this done soon.

**Comment by Motorrat** [comment]
@leobezerra you may want to consider the approach used in this extension too https://github.com/paypal/autosklearn-zeroconf

**Comment by leobezerra** [comment]
Hi guys.

A student of mine got an accuracy of ~98.4 on MNIST running for only 1h. The only thing he did was flattening the data beforehand.

Does that match your benchmarking?

**Comment by mfeurer** [comment]
Hi, could you explain what you mean by flattening? And no, we didn't achieve an accuracy of ~98.4 on MNIST within 1h (assuming that you're talking about test set accuracy). It would be great if your student could share his/her code for this.
